README.md
# README

[![NPM version][npm-badge]][npm-url]
[![Build Status][travis-badge]][travis-url]
[![Coverage Status][coveralls-badge]][coveralls-url]
[![Gittip][gittip-image]][gittip-url]
[![NPM downloads][npm-downloads]][npm-url]


[npm-badge]: https://img.shields.io/npm/v/@hotoo/orderby.svg?style=flat
[npm-url]: https://www.npmjs.com/package/@hotoo/orderby
[npm-downloads]: http://img.shields.io/npm/dm/@hotoo/orderby.svg?style=flat
[travis-badge]: https://travis-ci.org/hotoo/orderby.svg?branch=master
[travis-url]: https://travis-ci.org/hotoo/orderby
[coveralls-badge]: https://coveralls.io/repos/hotoo/orderby/badge.svg?branch=master
[coveralls-url]: https://coveralls.io/r/hotoo/orderby
[gittip-image]: https://img.shields.io/gittip/hotoo.svg?style=flat-square
[gittip-url]: https://www.gittip.com/hotoo/


SQL like syntax (ORDER BY) general sorting algorithm.

## USAGE

```js
const order = require('@hotoo/orderby');

[
 { a: 2, b: 'b' },
 { a: 1, b: 'b' },
 { a: 2, b: 'a' },
]
.sort(order.by('a', 'b DESC'));
```

return sorted array:

```js
[
 { a: 1, b: 'b' },
 { a: 2, b: 'b' },
 { a: 2, b: 'a' },
]
```


CHANGELOG.md
# CHANGELOG

## v2.0.1 (2020-01-09)

- feat: undefined > null > otherValues

## v2.0.0 (2020-01-04)

- 实现了完整的基本类型比较。

## v1.0.0 (2017-06-01)

- 实现简单的字符串和数值类型比较，其他类型默认转为字符串类型比较。
  日期类型被转成 GMT 格式（星期在最前面）导致排序错误。


test/util.test.js
'use strict';

require('should');
const {
  isBoolean,
  isDate,
  isFunction,
  isNull,
  isNumber,
  isObject,
  isRegExp,
  isString,
  isUndefined,
} = require('../util');
const util = require('../util');

describe('test/util.test', function() {
  const TYPES = [
    'Boolean',
    'Date',
    'Function',
    'Null',
    'Number',
    'Object',
    'RegExp',
    'String',
    'Undefined',
  ];
  const CASES = [
    [ 'Boolean', true ],
    [ 'Boolean', false ],
    [ 'Boolean', new Boolean(true) ],
    [ 'Boolean', new Boolean(false) ],
    [ 'Date', new Date() ],
    [ 'Date', new Date('2020-01-01T00:00:00.000Z') ],
    [ 'Date', new Date('Invalid Date') ],
    [ 'Function', function(){} ],
    [ 'Function', new Function('return false') ],
    [ 'Function', () => false ],
    [ 'Null', null ],
    [ 'Number', 0 ],
    [ 'Number', 1 ],
    [ 'Number', NaN ],
    [ 'Number', new Number(0) ],
    [ 'Number', new Number(1) ],
    [ 'Object', {} ],
    [ 'Object', { n: 1 } ],
    [ 'Object', new Object({}) ],
    [ 'Object', new Object({ newObject: 1 }) ],
    [ 'RegExp', /re/g ],
    [ 'RegExp', new RegExp('new RegExp', 'g') ],
    [ 'String', 'string' ],
    [ 'String', new String('new String') ],
    [ 'Undefined', undefined ],
  ];
  CASES.forEach(function(kase) {
    it(`is${kase[0]}(${String(kase[1])})`, function() {
      util['is' + kase[0]](kase[1]).should.eql(true);
    });
  });
  TYPES.forEach(function(type) {
    CASES
    .filter(kase => kase[0] !== type)
    .forEach(function(kase) {
      it(`NOT is${type}(${String(kase[1])})`, function() {
        util['is' + type](kase[1]).should.eql(false);
      });
    });
  });

});


test/compare.test.js
'use strict';

require('should');
const compare = require('../compare');

describe('test/compare.test', function() {
  const now = new Date();
  const CASES = [
    [ 'Number', 0, 0, 0 ],
    [ 'Number', 0, 1, -1 ],
    [ 'Number', 1, 0, 1 ],
    [ 'Number', 0, 100, -1 ],
    [ 'Number', 100, 0, 1 ],
    [ 'Number', NaN, NaN, 0 ],
    [ 'Number', NaN, 0, 1 ],
    [ 'Number', NaN, -1, 1 ],
    [ 'Number', NaN, 1, 1 ],
    [ 'Number', 0, NaN, -1 ],
    [ 'Number', -1, NaN, -1 ],
    [ 'Number', 1, NaN, -1 ],

    [ 'String', 'a', 'a', 0 ],
    [ 'String', '0', '0', 0 ],
    [ 'String', 'A', 'A', 0 ],
    [ 'String', '中', '中', 0 ],
    [ 'String', 'a', 'b', -1 ],
    [ 'String', 'b', 'a', 1 ],
    [ 'String', 'A', 'a', -1 ],
    [ 'String', 'a', 'A', 1 ],
    [ 'String', 'a', '中', -1 ],
    [ 'String', 'A', '中', -1 ],
    [ 'String', 'A', '中', -1 ],
    [ 'String', '中', '国', -1 ],
    [ 'String', '国', '中', 1 ],

    [ 'String', 'abcdefg', 'abcdefg', 0],
    [ 'String', '012345', '012345', 0],
    [ 'String', 'ABCDEFG', 'ABCDEFG', 0],
    [ 'String', new String('ABCDEFG'), 'ABCDEFG', 0],
    [ 'String', 'ABCDEFG', new String('ABCDEFG'), 0],
    [ 'String', new String('ABCDEFG'), new String('ABCDEFG'), 0],
    [ 'String', 'abcdefg', 'a', 1],
    [ 'String', 'a', 'abcdefg', -1],
    [ 'String', 'abcdefg', 'b', -1],
    [ 'String', 'b', 'abcdefg', 1],
    [ 'String', 'abcdefg', 'A', 1],
    [ 'String', 'A', 'abcdefg', -1],
    [ 'String', 'ABCDEFG', 'abcdefg', -1],
    [ 'String', 'abcdefg', 'ABCDEFG', 1],

    [ 'Boolean', true, true, 0],
    [ 'Boolean', false, false, 0],
    [ 'Boolean', true, false, 1],
    [ 'Boolean', false, true, -1],
    [ 'Boolean', new Boolean(true), new Boolean(true), 0],
    [ 'Boolean', new Boolean(false), new Boolean(false), 0],
    [ 'Boolean', new Boolean(true), new Boolean(false), 1],
    [ 'Boolean', new Boolean(false), new Boolean(true), -1],

    [ 'Date', now, now, 0],
    [ 'Date', new Date(2020,1,1), new Date(2020,1,1), 0],
    [ 'Date', new Date('2019-12-31'), new Date('2019-12-31'), 0],
    [ 'Date', new Date('2019-12-31T00:00:00.000Z'), new Date('2019-12-31T00:00:00.000Z'), 0],
    [ 'Date', new Date('2019-12-31T00:00:00.000Z'), new Date('2019-12-31T00:00:01.000Z'), -1],
    [ 'Date', new Date('2019-12-31T00:00:01.000Z'), new Date('2019-12-31T00:00:00.000Z'), 1],

    [ 'RegExp', /re/g, /re/g, 0],
    [ 'RegExp', /re/, /re/, 0],
    [ 'RegExp', /re/, /re/i, -1],
    [ 'RegExp', /re/, /re/m, -1],
    [ 'RegExp', /re/, /re/g, -1],
    [ 'RegExp', /ab/g, /re/g, -1],

    [ 'Function', function(){return false;}, function(){return true}, 0],
    [ 'Function', () => false, () => true, 0],
    [ 'Function', new Function('return false'), new Function('return true'), 0],

    [ 'Object', {}, {}, 0],
    [ 'Object', { n: 1 }, { n: 2 }, 0],

    [ 'MIX Date', new Date(), null, -1],
    [ 'MIX Date', null, new Date(), 1],

    [ 'Null', null, null, 0],
    [ 'MIX Null', null, undefined, -1],
    [ 'MIX Null', undefined, null, 1],
    [ 'MIX Null', null, 0, 1],
    [ 'MIX Null', null, 1, 1],
    [ 'MIX Null', null, '1', 1],
    [ 'MIX Null', null, 'a', 1],
    [ 'MIX Null', null, 'abcdefg', 1],
    [ 'MIX Null', null, true, 1],
    [ 'MIX Null', null, false, 1],
    [ 'MIX Null', null, new Date(), 1],
    [ 'MIX Null', null, /re/g, 1],
    [ 'MIX Null', null, function(){}, 1],

    [ 'MIX Null', 0, null, -1],
    [ 'MIX Null', 1, null, -1],
    [ 'MIX Null', '1', null, -1],
    [ 'MIX Null', 'a', null, -1],
    [ 'MIX Null', 'abcdefg', null, -1],
    [ 'MIX Null', true, null, -1],
    [ 'MIX Null', false, null, -1],
    [ 'MIX Null', new Date(), null, -1],
    [ 'MIX Null', /re/g, null, -1],
    [ 'MIX Null', function(){}, null, -1],

    [ 'MIX Undefined', undefined, undefined, 0],
    [ 'MIX Undefined', undefined, null, 1],
    [ 'MIX Undefined', undefined, 0, 1],
    [ 'MIX Undefined', undefined, 1, 1],
    [ 'MIX Undefined', undefined, '1', 1],
    [ 'MIX Undefined', undefined, 'a', 1],
    [ 'MIX Undefined', undefined, 'abcdefg', 1],
    [ 'MIX Undefined', undefined, new Date(), 1],
  ];

  describe('compare', function() {
    CASES.forEach(function(kase) {
      it(`${kase[0]}: copmare(${kase[1]}, ${kase[2]}) === ${kase[3]}`, function() {
        compare(kase[1], kase[2]).should.eql(kase[3]);
      });
    });
  });

  describe('compareStrings()', function() {
    it('compare same char', function() {
      const c = compare.compareStrings('a', 'a');
      c.should.equal(0);
    });
    it('compare different char to -1', function() {
      const c = compare.compareStrings('a', 'b');
      c.should.equal(-1);
    });
    it('compare different char to 1', function() {
      const c = compare.compareStrings('b', 'a');
      c.should.equal(1);
    });
    it('compare same string', function() {
      const c = compare.compareStrings('abcdefg', 'abcdefg');
      c.should.equal(0);
    });
    it('compare same string and new String', function() {
      const c = compare.compareStrings(new String('abcdefg'), 'abcdefg');
      c.should.equal(0);
    });
    it('compare same new String and string', function() {
      const c = compare.compareStrings('abcdefg', new String('abcdefg'));
      c.should.equal(0);
    });
    it('compare same new String', function() {
      const c = compare.compareStrings(new String('abcdefg'), new String('abcdefg'));
      c.should.equal(0);
    });
    it('compare long string and short sting to -1', function() {
      const c = compare.compareStrings('abcdefg', 'b');
      c.should.equal(-1);
    });
    it('compare long string and short string to 1', function() {
      const c = compare.compareStrings('abcdefg', 'aa');
      c.should.equal(1);
    });
    it('compare short string and long sting to -1', function() {
      const c = compare.compareStrings('a', 'abcdefg');
      c.should.equal(-1);
    });
    it('compare short string and long string to 1', function() {
      const c = compare.compareStrings('b', 'abcdefg');
      c.should.equal(1);
    });
  });
});


test/order_by.test.js
'use strict';

const order = require('../index');
require('should');

describe('test/order_by.test.js', () => {

  describe('order.by() simple', () => {
    it('order by number', () => {
      const list = [
        { n: 1 },
        { n: 3 },
        { n: 2 },
      ];
      const result = list.sort(order.by('n'));
      result.should.eql([
        { n: 1 },
        { n: 2 },
        { n: 3 },
      ]);
    });

    it('order by number asc', () => {
      const list = [
        { n: 1 },
        { n: 3 },
        { n: 2 },
      ];
      const result = list.sort(order.by('n asc'));
      result.should.eql([
        { n: 1 },
        { n: 2 },
        { n: 3 },
      ]);
    });

    it('order by number desc', () => {
      const list = [
        { n: 1 },
        { n: 3 },
        { n: 2 },
      ];
      const result = list.sort(order.by('n desc'));
      result.should.eql([
        { n: 3 },
        { n: 2 },
        { n: 1 },
      ]);
    });

    it('order by string', () => {
      const list = [
        { s: '1' },
        { s: '3' },
        { s: '2' },
      ];
      const result = list.sort(order.by('s'));
      result.should.eql([
        { s: '1' },
        { s: '2' },
        { s: '3' },
      ]);
    });

    it('order by string asc', () => {
      const list = [
        { s: '1' },
        { s: '3' },
        { s: '2' },
      ];
      const result = list.sort(order.by('s asc'));
      result.should.eql([
        { s: '1' },
        { s: '2' },
        { s: '3' },
      ]);
    });

    it('order by string desc', () => {
      const list = [
        { s: '1' },
        { s: '3' },
        { s: '2' },
      ];
      const result = list.sort(order.by('s desc'));
      result.should.eql([
        { s: '3' },
        { s: '2' },
        { s: '1' },
      ]);
    });
  });

  describe('order.by() more', () => {
    it('order by more number', () => {
      const list = [
        { n: 1, n2: 3 },
        { n: 3, n2: 2 },
        { n: 2, n2: 1 },
        { n: 2, n2: 3 },
        { n: 2, n2: 2 },
      ];
      const result = list.sort(order.by('n', 'n2'));
      result.should.eql([
        { n: 1, n2: 3 },
        { n: 2, n2: 1 },
        { n: 2, n2: 2 },
        { n: 2, n2: 3 },
        { n: 3, n2: 2 },
      ]);
    });

    it('order by more number asc', () => {
      const list = [
        { n: 1, n2: 3 },
        { n: 3, n2: 2 },
        { n: 2, n2: 1 },
        { n: 2, n2: 3 },
        { n: 2, n2: 2 },
      ];
      const result = list.sort(order.by('n asc', 'n2 asc'));
      result.should.eql([
        { n: 1, n2: 3 },
        { n: 2, n2: 1 },
        { n: 2, n2: 2 },
        { n: 2, n2: 3 },
        { n: 3, n2: 2 },
      ]);
    });

    it('order by more number desc', () => {
      const list = [
        { n: 1, n2: 3 },
        { n: 3, n2: 2 },
        { n: 2, n2: 1 },
        { n: 2, n2: 3 },
        { n: 2, n2: 2 },
      ];
      const result = list.sort(order.by('n desc', 'n2 desc'));
      result.should.eql([
        { n: 3, n2: 2 },
        { n: 2, n2: 3 },
        { n: 2, n2: 2 },
        { n: 2, n2: 1 },
        { n: 1, n2: 3 },
      ]);
    });

    it('order by more string', () => {
      const list = [
        { s: '1', s2: '3' },
        { s: '3', s2: '2' },
        { s: '2', s2: '1' },
        { s: '2', s2: '3' },
        { s: '2', s2: '2' },
      ];
      const result = list.sort(order.by('s', 's2'));
      result.should.eql([
        { s: '1', s2: '3' },
        { s: '2', s2: '1' },
        { s: '2', s2: '2' },
        { s: '2', s2: '3' },
        { s: '3', s2: '2' },
      ]);
    });

    it('order by more string asc', () => {
      const list = [
        { s: '1', s2: '3' },
        { s: '3', s2: '2' },
        { s: '2', s2: '1' },
        { s: '2', s2: '3' },
        { s: '2', s2: '2' },
      ];
      const result = list.sort(order.by('s asc', 's2 asc'));
      result.should.eql([
        { s: '1', s2: '3' },
        { s: '2', s2: '1' },
        { s: '2', s2: '2' },
        { s: '2', s2: '3' },
        { s: '3', s2: '2' },
      ]);
    });

    it('order by more string desc', () => {
      const list = [
        { s: '1', s2: '3' },
        { s: '3', s2: '2' },
        { s: '2', s2: '1' },
        { s: '2', s2: '3' },
        { s: '2', s2: '2' },
      ];
      const result = list.sort(order.by('s desc', 's2 desc'));
      result.should.eql([
        { s: '3', s2: '2' },
        { s: '2', s2: '3' },
        { s: '2', s2: '2' },
        { s: '2', s2: '1' },
        { s: '1', s2: '3' },
      ]);
    });

    it('order by same list', () => {
      const list = [
        { s: '1', s2: '2' },
        { s: '1', s2: '2' },
        { s: '1', s2: '2' },
        { s: '1', s2: '2' },
        { s: '1', s2: '2' },
      ];
      const result = list.sort(order.by('s desc', 's2 desc'));
      result.should.eql([
        { s: '1', s2: '2' },
        { s: '1', s2: '2' },
        { s: '1', s2: '2' },
        { s: '1', s2: '2' },
        { s: '1', s2: '2' },
      ]);
    });
  });

  describe('order.by() complex', () => {
    it('order by complex number and string', () => {
      const list = [
        { n: 1, s: '3' },
        { n: 3, s: '2' },
        { n: 2, s: '1' },
        { n: 2, s: '3' },
        { n: 2, s: '2' },
      ];
      const result = list.sort(order.by('n', 's'));
      result.should.eql([
        { n: 1, s: '3' },
        { n: 2, s: '1' },
        { n: 2, s: '2' },
        { n: 2, s: '3' },
        { n: 3, s: '2' },
      ]);
    });

    it('order by complex number and string asc', () => {
      const list = [
        { n: 1, s: '3' },
        { n: 3, s: '2' },
        { n: 2, s: '1' },
        { n: 2, s: '3' },
        { n: 2, s: '2' },
      ];
      const result = list.sort(order.by('n asc', 's asc'));
      result.should.eql([
        { n: 1, s: '3' },
        { n: 2, s: '1' },
        { n: 2, s: '2' },
        { n: 2, s: '3' },
        { n: 3, s: '2' },
      ]);
    });

    it('order by complex number and string desc', () => {
      const list = [
        { n: 1, s: '3' },
        { n: 3, s: '2' },
        { n: 2, s: '1' },
        { n: 2, s: '3' },
        { n: 2, s: '2' },
      ];
      const result = list.sort(order.by('n desc', 's desc'));
      result.should.eql([
        { n: 3, s: '2' },
        { n: 2, s: '3' },
        { n: 2, s: '2' },
        { n: 2, s: '1' },
        { n: 1, s: '3' },
      ]);
    });

    it('order by complex number and string desc and null', () => {
      const list = [
        { n: 1, s: '3', date: '2019-12-17T04:04:20.000Z', nil: null },
        { n: 3, s: '2', date: '2019-12-26T07:36:57.000Z', nil: null },
        { n: 2, s: '1', date: '2019-12-19T09:05:20.000Z', nil: null },
        { n: 3, s: '1', date: '2019-12-19T09:05:20.000Z', nil: null },
        { n: 2, s: '3', date: '2019-12-18T09:05:20.000Z', nil: 2 },
        { n: 2, s: '2', date: '2019-12-30T09:05:20.000Z', nil: 1 },
      ];
      const result = list.sort(order.by('nil', 'date desc', 'n'));
      result.should.eql([
        { n: 2, s: '2', date: '2019-12-30T09:05:20.000Z', nil: 1 },
        { n: 2, s: '3', date: '2019-12-18T09:05:20.000Z', nil: 2 },
        { n: 3, s: '2', date: '2019-12-26T07:36:57.000Z', nil: null },
        { n: 2, s: '1', date: '2019-12-19T09:05:20.000Z', nil: null },
        { n: 3, s: '1', date: '2019-12-19T09:05:20.000Z', nil: null },
        { n: 1, s: '3', date: '2019-12-17T04:04:20.000Z', nil: null },
      ]);
    });

    it('order by complex real', () => {
      const list = [
        {
          'activityId': 'S-20191212',
          'gameStartDate': '2019-12-12T07:00:00.000Z',
          'orderRankNo': null,
          'gmtCreate': '2019-12-26T07:36:57.000Z', // 真皮
        },
        {
          'activityId': 'S-20191212',
          'gameStartDate': '2019-12-12T07:00:00.000Z',
          'orderRankNo': null,
          'gmtCreate': '2019-12-19T09:05:20.000Z',
        },
        {
          'activityId': 'S-20191212',
          'gameStartDate': '2019-12-12T07:00:00.000Z',
          'orderRankNo': null,
          'gmtCreate': '2019-12-17T04:04:20.000Z',
        },
      ];
      const result = list.sort(order.by('gameStartDate DESC', 'activityId', 'orderRankNo', 'gmtCreate DESC'));
      result.should.eql([
        {
          'activityId': 'S-20191212',
          'gameStartDate': '2019-12-12T07:00:00.000Z',
          'orderRankNo': null,
          'gmtCreate': '2019-12-26T07:36:57.000Z', // 真皮
        },
        {
          'activityId': 'S-20191212',
          'gameStartDate': '2019-12-12T07:00:00.000Z',
          'orderRankNo': null,
          'gmtCreate': '2019-12-19T09:05:20.000Z',
        },
        {
          'activityId': 'S-20191212',
          'gameStartDate': '2019-12-12T07:00:00.000Z',
          'orderRankNo': null,
          'gmtCreate': '2019-12-17T04:04:20.000Z',
        },
      ]);
    });
  });

  describe('order.by RULE ERROR', () => {
    it('rule error', () => {
      const list = [
        { n: 1 },
        { n: 3 },
        { n: 2 },
      ];
      let err = 0;
      try {
        const result = list.sort(order.by('n ABC'));
        err = 1;
      } catch(ex) {
        err = 2;
        ex.message.should.equal('[order.by] rule error: n ABC');
      }
      err.should.equal(2);
    });
  });

  describe('order.by NOT FOUND CLOUMN NAME', () => {
    it('cloumn name error', () => {
      const list = [
        { n: 1 },
        { n: 3 },
        { n: 2 },
      ];
      let err = 0;
      try {
        const result = list.sort(order.by('__NOT_EXIST__'));
        err = 1;
      } catch(ex) {
        err = 2;
        ex.message.should.equal('[order.by] not found "__NOT_EXIST__" in data');
      }
      err.should.equal(2);
    });
  });
});


compare.js
'use strict';
const {
  isBoolean,
  isDate,
  isFunction,
  isNull,
  isNumber,
  isObject,
  isRegExp,
  isString,
  isUndefined,
} = require('./util');

/**
 * 一维排序算法。
 * @param {Object} a 对比对象
 * @param {Object} b 对比对象
 * @return {Number} 如果 a 要排在 b 之前，返回 -1
 *                  如果 a 要排在 b 之后，返回 1
 *                  如果 a 和 b 相等，返回 0
 *                  特殊的，对于有 null 或 undefined 的情况：
 *
 *                  任何有效值 < null < undefined
 *
 *                  如果你需要将某些反向设计的特殊值（如 -1，0 等）
 *                  排到正常值（如1，2等）之后，可以将特殊值处理为 null
 */
function compare(a, b) {
  if (isUndefined(a) || isUndefined(b) || isNull(a) || isNull(b)) {
    const la = specialValue(a);
    const lb = specialValue(b);
    return compareNumbers(la, lb);
  }

  // 以下为 a 和 b 均不为 null 或 undefined。

  if (isRegExp(a) && isRegExp(b)) {
    return compareRegExps(a, b);
  }
  // 不可比较
  if (isObject(a) && isObject(b)) {
    return 0;
  }
  if (isFunction(a) && isFunction(b)) {
    return 0;
  }

  // 数值比较
  if (isNumber(a) && isNumber(b)) {
    return compareNumbers(a, b);
  }
  if (isDate(a) && isDate(b)) {
    const da = a.getTime();
    const db = b.getTime();
    return compareNumbers(da, db);
  }
  if(isBoolean(a) && isBoolean(b)) {
    return compareNumbers(Number(a), Number(b));
  }

  // 默认使用字符串比较
  return compareStrings(String(a), String(b));
}

/**
 * 将特殊值（undefined, null）转换为数值进行排序。
 * undefined < null < value
 * @param {Object} value 需要处理的特殊值。
 * @return {Number}
 */
function specialValue(value) {
  if (isUndefined(value)) return 2;
  if (isNull(value)) return 1;
  return 0;
}

/**
 * 比较数值。
 *
 * @param {Number} a 比较对象 a
 * @param {Number} b 比较对象 b
 * @return {Number} 比较结果。
 */
function compareNumbers(a, b) {
  if (isNaN(a) && isNaN(b)) {
    return 0
  }
  if (isNaN(a)) {
    return 1;
  }
  if (isNaN(b)) {
    return -1;
  }
  if (a > b) {
    return 1;
  }
  if (a < b) {
    return -1;
  }
  return 0;
}

function compareStrings(a, b) {
  if (a === b) { return 0; }
  const maxLength = Math.max(a.length, b.length);
  for (let i = 0; i < maxLength; i++) {
    let ca = a.charCodeAt(i) || -1;
    let cb = b.charCodeAt(i) || -1;
    if (ca > cb) {
      return 1;
    }
    if (ca < cb) {
      return -1;
    }
  }
  return 0;

  // "A".charCodeAt(0) === 65
  // "a".charCodeAt(0) === 97
  //
  // 且
  // "A" < "a" === true
  //
  // 但
  // "A".localeCompare('a') === 1
  // 用 localeCompare 时 A > a，不符合预期
  //! return a.localeCompare(b);
}

function regexp2string(reg) {
  return '/' + reg.source + '/' +
    (reg.ignoreCase ? 'i' : '') +
    (reg.multiline ? 'm' : '') +
    (reg.global ? 'g' : '');
}

function compareRegExps(a, b) {
  const ra = regexp2string(a);
  const rb = regexp2string(b);
  return compareStrings(ra, rb);
}

module.exports = compare;
module.exports.compareStrings = compareStrings;
module.exports.compareNumbers = compareNumbers;
module.exports.compareRegExps = compareRegExps;


index.js
'use strict';
/**
 * 类似 SQL 的 ORDER BY 通用排序算法。
 *
 * const order = require('./order');
 *
 * [
 *  {a:2, b:'b'},
 *  {a:1, b:'a'},
 *  {a:2, b:'a'},
 * ].sort(order.by('a', 'b DESC'));
 *
 * @author 冒顿
 * @version 2017-05-31
 */
const compare = require('./compare');
const { isUndefined } = require('./util');

const RE_ORDER_BY_RULE = /^([^\s]+)(?:\s+(ASC|DESC))?$/i;

/**
 * list.sort(orderBy('columnA DESC','columnB'))
 * @return {Number} -1, 0, 1
 */
function orderBy() {
  const rules = [];
  for (let i = 0, l = arguments.length; i < l; i++) {
    const rule = arguments[i];
    const m = RE_ORDER_BY_RULE.exec(rule);
    if (!m || !m[1]) {
      throw new Error('[order.by] rule error: ' + rule);
    }
    rules.push([
      m[1],
      (m[2] || 'ASC').toUpperCase(),
    ]);
  }
  return (a, b) => {
    for (let i = 0, l = rules.length; i < l; i++) {
      const rule = rules[i];
      const columnName = rule[0];
      let order;
      if (isUndefined(a[columnName]) || isUndefined(b[columnName])) {
        throw new Error('[order.by] not found "' + columnName + '" in data');
      }
      if (rule[1] === 'ASC') {
        order = compare(a[columnName], b[columnName]);
      } else { // DESC
        order = compare(b[columnName], a[columnName]);
      }
      if (order !== 0) {
        return order;
      }
    }
    return 0;
  };
}

module.exports = orderBy;
module.exports.by = orderBy;
